### PR TITLE
Add end-to-end integration test for video test app

### PR DESCRIPTION
**Pull request by ldai1**
**Issue #:**

**Description of changes:**
Add end-to-end test for video test app. 
Tests cover apis:
1. addVideoTile 
2. startLocalVideoTile
3. hasStartedLocalVideoTile
4. haveVideoTilesWithStreams
5. haveVideoTileForAttendeeId
6. getLocalVideoTile
7. getAllVideoTiles
8. getAllRemoteVideoTiles 
9. removeAllVideoTiles
10. stopLocalVideoTile 
11. sendTileStateUpdate

Test cover scenarios:
    * 2 attendees join from 2 different browser sessions
    * Both attendees turn on their local video
    * Verify that both attendees can see their local video on
    * Verify that both attendees can see the other attendee’s remote video
    * Both attendees turn off their video
    * Verify that both attendees’ local videos off
    * Verify that both attendees’ remote videos off

**Testing**

1. Have you successfully run `npm run build:release` locally? yes
2. How did you test these changes? manullly and run the chromDrive
3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it? Video_test
4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved? No
5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved? No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: integration/js/VideoEnd2EndTest.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Both attendees turn off their video
+    await ClickStopLocalVideoButton.executeStep(this, session);
+    await WaitForAllParticipantsToTurnVideoOff.executeStep(this, session, '1');
+
+    // Verify that both attendees’ local videos off
+    await ClickHasStartedLocalVideoTileButton.executeStep(this, session);
+    await HasStartedLocalVideoTileCheck.executeStep(this, session, 'false');
+
+    await TestUtils.waitAround(1000);
```
Comment (michhyun1) [inline integration/js/VideoEnd2EndTest.js:52]: Why is this necessary?

File: integration/js/steps/WaitForAllParticipantsToTurnVideoOff.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    const step = new WaitForAllParticipantsToTurnVideoOff(KiteBaseTest, sessionInfo, remoteParticipantCount);
+    await step.execute(KiteBaseTest);
+  }
+
+  stepDescription() {
+    return 'Waiting for other participants to turn off their videos';
+  }
+
+  waitCompleteCondition() {
+    return this.test.numVideoRemoteOff.toString() === this.remoteParticipantCount;
```
Comment (michhyun1) [inline integration/js/steps/WaitForAllParticipantsToTurnVideoOff.js:20]: Can we re-use 'WaitForRemoteParticipantsToTurnVideoOff' ?

File: integration/js/steps/WaitForAllParticipantsToJoinMeeting.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  stepDescription() {
+    return 'Waiting for all participants to join the meeting';
+  }
+
+  waitCompleteCondition() {
+    console.log("participantCount " + this.participantCount);
+    console.log("numberOfParticipant " + this.numberOfParticipant);
+    return this.numberOfParticipant.toString() === this.participantCount;
```
Comment (michhyun1) [inline integration/js/steps/WaitForAllParticipantsToJoinMeeting.js:21]: I think we aready have a Wait step defined for this? Can we reuse that?

File: integration/js/steps/WaitForAllParticipantsToTurnVideoOn.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    const step = new WaitForAllParticipantsToTurnVideoOn(KiteBaseTest, sessionInfo, remoteParticipantCount);
+    await step.execute(KiteBaseTest);
+  }
+
+  stepDescription() {
+    return 'Waiting for all participants to turn on their audio';
+  }
+
+  waitCompleteCondition() {
+    return this.test.numVideoRemoteOn.toString() === this.remoteParticipantCount
```
Comment (michhyun1) [inline integration/js/steps/WaitForAllParticipantsToTurnVideoOn.js:20]: Can we re-use "WaitForRemoteParticipantsToTurnVideoOn"?

File: integration/js/VideoEnd2EndTest.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    super(name, kiteConfig, "TestAppVideoEnd2End");
+  }
+
+  async runIntegrationTest() {
+    const attendeeName = uuidv4();
+    const session = this.seleniumSessions[0];
+    const useSimulcast = this.useSimulcast;
+    await WaitForMeetingToBeCreated.executeStep(this, session);
+    await OpenAppStep.executeStep(this, session);
+    await JoinVideoTestMeetingStep.executeStep(this, session, attendeeName, this.meetingTitle, useSimulcast);
```
Comment (michhyun1) [inline integration/js/VideoEnd2EndTest.js:18]: I think you can just create the attendeeName inside this step, unless you are using it somewhere else.

File: integration/js/app/video_test/video_test.ts [hunk 6]
```diff
@@ -113,6 +113,89 @@ export class DemoMeetingApp implements AudioVideoObserver {
...
+      remoteVideoTilesCountObject.innerText = JSON.stringify(totalRemoteVideoTilesCount, null, '  ');;
+      getAllRemoteVideoTilesForm.reset();
+    });
+
+
+    const haveVideoTileForAttendeeIdForm = document.getElementById('have-video-tile-for-attendeeId-form') as HTMLFormElement;
+    haveVideoTileForAttendeeIdForm.addEventListener('submit', e => {
+      e.preventDefault();
+      const attendeeIdInput = document.getElementById('have-video-tile-for-attendee-id') as HTMLInputElement;
+      console.log("attendeeIdInput: " + attendeeIdInput.value);
```
[the remark this answers is not in the document]
Comment (ldai1) [reply integration/js/app/video_test/video_test.ts:170]: Good catch. Will remove

File: integration/js/VideoEnd2EndTest.js (hunk 1 above)
Comment (ldai1) [reply integration/js/VideoEnd2EndTest.js:52]: Due to race condition from two browers to turn off local video, the GetAllRemoteVideoTiles step might still return 1. So I put a wait time here to make sure both users have turned off their video so that GetAllRemoteVideoTiles step will return 0.

File: integration/js/steps/WaitForAllParticipantsToJoinMeeting.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  waitCompleteMessage() {
+    return "all all participants present"
```
[the remark this answers is not in the document]
Comment (ldai1) [reply integration/js/steps/WaitForAllParticipantsToJoinMeeting.js:25]: Good catch. Will change.

File: integration/js/steps/WaitForAllParticipantsToTurnVideoOff.js (hunk 2 above)
Comment (ldai1) [reply integration/js/steps/WaitForAllParticipantsToTurnVideoOff.js:20]: I also wanted to re-use the WaitForRemoteParticipantsToTurnVideoOff but it turns out to be the waitCompleteCondition can't be met `(this.test.numVideoRemoteOff.toString() === this.numberOfParticipant )` In order not to break any existing tests, I add this new step to compare the `this.test.numVideoRemoteOff.toString() ` and the `this.remoteParticipantCount` In my case, the remoteParticipantCount is 1 and the numVideoRemoteOff is also 1.

File: integration/js/steps/WaitForAllParticipantsToJoinMeeting.js (hunk 3 above)
Comment (ldai1) [reply integration/js/steps/WaitForAllParticipantsToJoinMeeting.js:21]: yes, we already have a WaitForRemoteParticipantsToJoinMeeting, but somehow the waitCompleteCondition can not meet: `this.test.numRemoteJoined.toString() === this.numberOfParticipant`. In my case, the numRemoteJoined is 1 and the numberOfParticipant is 2. In order not to break any existing test that uses the WaitForRemoteParticipantsToJoinMeeting, I have to create a new step called WaitForAllparticipantsToJoinMeeting.

File: integration/js/steps/WaitForAllParticipantsToTurnVideoOn.js (hunk 4 above)
Comment (ldai1) [reply integration/js/steps/WaitForAllParticipantsToTurnVideoOn.js:20]: Same reason of not using WaitForParticipantsToTurnVideoOff. The waitCompleteCondition cannot meet: `this.test.numVideoRemoteOn.toString() === this.remoteParticipantCount`

File: integration/js/VideoEnd2EndTest.js (hunk 5 above)
Comment (ldai1) [reply integration/js/VideoEnd2EndTest.js:18]: You are right, I don't need to create the attendeeName by myself. Will remove it.

File: integration/js/VideoEnd2EndTest.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    await ClickStartLocalVideoButton.executeStep(this, session);
+    await WaitForAllParticipantsToTurnVideoOn.executeStep(this, session, '1');
+
+    // Verify that both attendees can see their local video on
+    await ClickHasStartedLocalVideoTileButton.executeStep(this, session);
+    await HasStartedLocalVideoTileCheck.executeStep(this, session, 'true');
+    await ClickGetLocalVideoTileButton.executeStep(this, session);
+    await GetLocalVideoTileCheck.executeStep(this, session, '1')
+   
+    await ClickHaveVideoTileForAttendeeIdButton.executeStep(this, session);
```
[the remark this answers is not in the document]
Comment (ldai1) [reply integration/js/VideoEnd2EndTest.js:32]: Agree. Will make the change to directly enter the boundAttendeeId as an input.

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,15 @@ All notable changes to this project will be documented in this file.
...
 
+## [Unreleased]
+- Add End-to-end Integration test for Video Test App
+
+### Changed
+
+### Removed
+
+### Fixed
+
```
[the remark this answers is not in the document]
Comment (ldai1) [reply CHANGELOG.md]: Sure, I will add this change to the existing [Unreleased] section.

File: integration/js/steps/WaitForAllParticipantsToTurnVideoOff.js (hunk 2 above)
Comment (michhyun1) [reply integration/js/steps/WaitForAllParticipantsToTurnVideoOff.js:20]: It looks like the logic is the same for both steps:

this.test.numVideoRemoteOff.toString() === this.numberOfParticipant

this.test.numVideoRemoteOff.toString() === this.remoteParticipantCount;

This step should be reusable, it might take some adjusting on the original test

**Review by michhyun1 — APPROVED** [review]
